concordium-consensus/src/Concordium/MultiVersion.hs

```diff
@@ -559,11 +570,129 @@ makeMultiVersionRunner
         putMVar mvWriteLock ()
         startTransactionPurgingThread mvr
         return mvr
 
+
+-- |Start a consensus with a new genesis.
+-- It is assumed that the thread holds the write lock.
+-- This calls 'notifyRegenesis' to alert the P2P layer of the new genesis block so that p2p layer
+-- starts up with a correct list of genesis blocks.
+--
+-- Startup works as follows.
+--
+-- 1. Genesis data is partially decoded to determine the protocol version where
+-- the chain is supposed to start.
+-- 2. For each protocol version starting at the initial one we attempt to load
+-- the existing state for the protocol version and genesis index. This loading
+-- is minimal and no state caching is done. Only enough state is loaded so that
+-- queries for old blocks, data in blocks, are supported.
+-- 3. After this process we end up in one of two slightly different options.
+--    - Either there is no state at all. In that case we need to decode the
+--    supplied genesis data fully and start a chain with it. We start a new
+--    chain with 'newGenesis'.
+--    - Or we have loaded at least one state. In this case we must check whether
+--      the state already contains an effective protocol update or not.
+--      'checkForProtocolUpdate' is used for this and it might start a new chain.
+startupSkov ::
+    forall gsconf finconf.
+    ( MultiVersionStateConfig gsconf,
+      MultiVersion gsconf finconf,
+      SkovConfiguration gsconf finconf UpdateHandler
+    ) =>
+    -- |Genesis data, the decoder for it and the byte array representation. The
+    -- reason this does not take the genesis data directly is that it can be
+    -- expensive to deserialize, and if the database already exists then it is
+    -- not needed.
+    Either ByteString PVGenesisData ->
+    MVR gsconf finconf ()
+startupSkov genesis = do
+    initProtocolVersion <- case genesis of
+        Left genBS -> case runGet getPVGenesisDataPV genBS of
+            Left err -> throwM (InvalidGenesisData err)
+            Right spv -> return spv
+        Right (PVGenesisData (_ :: GenesisData pvOrig)) -> return (SomeProtocolVersion (protocolVersion @pvOrig))
+    let loop :: SomeProtocolVersion
+             -- ^Protocol version at which to attempt to load the state.
+             -> Maybe (EVersionedConfiguration gsconf finconf)
+             -- ^If this is the first iteration of the loop then this will be 'Nothing'. Otherwise it is the
+             -- versioned configuration produced in the previous iteration of the loop.
+             -> GenesisIndex
+             -- ^Genesis index at which to attempt to load the state.
+             -> AbsoluteBlockHeight
+             -- ^Absolute block height of the genesis block of the new chain.
+             -> MVR gsconf finconf ()
+        loop (SomeProtocolVersion (_ :: SProtocolVersion pv)) first vcIndex vcGenesisHeight = do
+            let comp = MVR $
```

Comment:
This function builds an `Either` value with quite a few `Maybe`s nested inside. Perhaps, using a custom type (with 5 constructors, if I counted right) could help to improve the readability of the pattern matching tree that consumes this value.

Reply:
I am in favor of this.

I have documented the branches quite explicitly which in my opinion documents the results of this function quite well.

If I added a new type then those comments would just be next to the type. Seeing that this is just an inline loop I do not see the advantage.